test/instance.spec.js

```diff
@@ -0,0 +1,58 @@
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+chai.use(require('dirty-chai'))
+chai.use(require('chai-spies'))
+const expect = chai.expect
+
+const PubsubBaseProtocol = require('../src')
+const { createPeerInfo, mockRegistrar } = require('./utils')
+
+describe('should validate instance parameters', () => {
+  let peerInfo
+
+  before(async () => {
+    peerInfo = await createPeerInfo()
+  })
+
+  it('should throw if no debugName is provided', () => {
+    expect(() => {
+      new PubsubBaseProtocol() // eslint-disable-line no-new
+    }).to.throw()
+  })
+
+  it('should throw if no multicodec is provided', () => {
+    expect(() => {
+      new PubsubBaseProtocol('pubsub') // eslint-disable-line no-new
+    }).to.throw()
+  })
+
+  it('should throw if no peerInfo is provided', () => {
+    expect(() => {
+      new PubsubBaseProtocol('pubsub', '/pubsub/1.0.0') // eslint-disable-line no-new
+    }).to.throw()
+  })
+
+  it('should throw if an invalid peerInfo is provided', () => {
+    expect(() => {
+      new PubsubBaseProtocol('pubsub', '/pubsub/1.0.0', 'fake-peer-info') // eslint-disable-line no-new
+    }).to.throw()
+  })
+
+  it('should throw if no registrar object is provided', () => {
+    expect(() => {
+      new PubsubBaseProtocol('pubsub', '/pubsub/1.0.0', peerInfo) // eslint-disable-line no-new
+    }).to.throw()
+  })
+
+  it('should accept valid parameters', () => {
+    expect(() => {
+      new PubsubBaseProtocol('pubsub', '/pubsub/1.0.0', peerInfo, mockRegistrar) // eslint-disable-line no-new
+    }).not.to.throw()
+  })
+})
+
+describe('should require functions to be implemented', () => {
+
```

Comment:
Are tests here planned? It's empty right now.

Reply:
Well, this was when I had some thoughts if we should extract a suite of tests that implementations should import and use, kind of like an interface. But I don't want to overcomplicate this now too. What to do you think?

Reply:
Ah, I think a follow up PR for a compliance test suite would be fine. No need to push it all into this 1 PR. An interface test suite would definitely be great to have.